test/Web/Rollbar/Test.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Web.Rollbar.Test where
+
+import Web.Rollbar
+    ( AccessToken(..)
+    , CodeVersion(..)
+    , Environment(..)
+    , Event(..)
+    , EventLevel(..)
+    , HasRollbarCfg(..)
+    , Host(..)
+    , RollbarCfg(..)
+    )
+import qualified Web.Rollbar.Internal as I
+
+import Control.Lens.TH (makeLenses)
+import Control.Monad.Reader (runReaderT)
+import Data.Aeson (ToJSON(..), defaultOptions, genericToEncoding)
+import Data.Aeson.Encode.Pretty (Config(..), Indent(..), defConfig, encodePretty')
+import Data.ByteString.Lazy (ByteString)
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import qualified Data.Text.IO as TIO
+import GHC.Generics (Generic)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Golden (goldenVsFile, goldenVsString, writeBinaryFile)
+
+newtype Context = Context
+    { _ctxRollbarCfg :: RollbarCfg
+    }
+
+makeLenses ''Context
+
+instance HasRollbarCfg Context where
+    rollbarCfg = ctxRollbarCfg
+
+---
+---
+---
+data SampleData = SampleData
+    { sampleFoo :: Bool
+    , sampleBar :: Integer
+    } deriving (Generic)
+
+instance ToJSON SampleData where
+    toEncoding = genericToEncoding defaultOptions
+
+minimalEvent :: Event
+minimalEvent =
+    Event
+    { _eventLevel = Info
+    , _eventUUID = Nothing
+    , _eventTitle = "title-testing"
+    , _eventMessage = "message-testing"
+    , _eventData = Nothing
+    , _eventContext = Nothing
+    }
+
+fullEvent :: Event
+fullEvent =
+    minimalEvent
+    { _eventLevel = Error
+    , _eventUUID = Just "deadbeef-dead-beef-dead-deadbeefdead"
+    , _eventData = Just (toJSON SampleData {sampleFoo = True, sampleBar = 42})
+    , _eventContext = Just "context-testing"
+    }
+
+minimalConfig :: RollbarCfg
+minimalConfig =
+    RollbarCfg
+    { _rollbarCfgToken = AccessToken "token-testing"
+    , _rollbarCfgEnvironment = Environment "environment-testing"
+    , _rollbarCfgHost = Nothing
+    , _rollbarCfgCodeVersion = Nothing
+    , _rollbarCfgMute = False
+    }
+
+fullConfig :: RollbarCfg
+fullConfig =
+    minimalConfig
+    { _rollbarCfgHost = Just (Host "host-testing")
+    , _rollbarCfgCodeVersion = Just (CodeVersion "code-version-testing")
+    , _rollbarCfgMute = False
+    }
+
+---
+---
+---
+test_encodeEvent :: TestTree
+test_encodeEvent =
+    let events = [("minimal", minimalEvent), ("full", fullEvent)]
+        configs = [("minimal", minimalConfig), ("full", fullConfig)]
+        tests :: [((String, String), (RollbarCfg, Event))]
+        tests =
+            [ ( (en <> " event with " <> cn <> " config", "event-" <> en <> "-config-" <> cn)
+              , (cv, ev))
+            | (en, ev) <- events
+            , (cn, cv) <- configs
+            ]
```

Comment:
I think this effect of Cartesian product can be neatly achieved via QuickCheck. I don't have an example though.

Also, I think it might be cleaner to not declare `tests` separately, but rather have it inline under `testGroup` below, kind lilke this:

```
testGroup "Encoding" $
    [ goldenVsString name fileName $ test cv ev
    | (en, ev) <- [("minimal", minimalEvent), ("full", fullEvent)]
    , (cn, cv) <- [("minimal", minimalConfig), ("full", fullConfig)]
    , let fileName = "test/golden/" <> "event-" <> en <> "-config-" <> cn <> ".json"
    , let name = en <> " event with " <> cn <> " config"
    ]
```

Reply:
Wow, that is so much nicer — thank you! I didn’t know one can have `let` bindings inside the list comprehension.

Re: QuickCheck. Good point. I’ve used it before but more for property testing where here it’s about snapshot testing against files which `Tasty.Golden` supports well. I think your simplification makes it really pleasant 🙇